tests/phpunit/includes/testcase-multisites.php

```diff
@@ -0,0 +1,312 @@
+			self::$blog_with_pll_default_links,
+			array( $this->languages['en'], $this->languages['fr'] ),
+			array(),
+			$this->plain_structure
+		);
+
+		// Set up blog with Polylang not activated and plain permalinks.
+		$this->set_up_blog_without_pll( self::$blog_without_pll_plain_links, $this->plain_structure );
+
+		// Set up blog with Polylang not activated and pretty permalinks.
+		$this->set_up_blog_without_pll( self::$blog_without_pll_pretty_links, $this->pretty_structure );
+	}
+
+	public function tear_down() {
+		restore_current_blog();
+
+		wp_delete_site( self::$blog_without_pll_pretty_links->blog_id );
+		wp_delete_site( self::$blog_with_pll_directory->blog_id );
+		wp_delete_site( self::$blog_with_pll_domains->blog_id );
+		wp_delete_site( self::$blog_with_pll_default_links->blog_id );
+
+		wp_update_network_site_counts();
+
+		parent::tear_down();
+	}
+
+	/**
+	 * Sets a blog up with the correct structure and activates Polylang's plugins.
+	 * Also creates languages.
+	 *
+	 * @global $wp_rewrite
+	 *
+	 * @param WP_Site $blog      Blog to set up.
+	 * @param array   $languages Languages to create.
+	 * @param array   $options   Polylang options to use.
+	 * @param string  $structure Permalink structure to use.
+	 * @return void
+	 */
+	protected function set_up_blog_with_pll( WP_Site $blog, array $languages, array $options, string $structure ) {
+		global $wp_rewrite;
+
+		switch_to_blog( $blog->blog_id );
+
+		$wp_rewrite->init();
+		$wp_rewrite->extra_rules_top = array();
+		$wp_rewrite->set_permalink_structure( $structure );
+
+		$plugins = get_option( 'active_plugins', array() );
+		update_option( 'active_plugins', array_merge( $plugins, $this->get_plugin_names() ) );
+
+		$pll_admin = $this->get_pll_admin_env( $options, false );
+
+		foreach ( $languages as $language ) {
+			$added = $pll_admin->model->add_language( $language );
+		}
+
+		$pll_admin->model->get_links_model()->init();
+		$pll_admin->init();
+		$wp_rewrite->flush_rules();
+
+		restore_current_blog();
+	}
+
+	/**
+	 * Sets a blog up with the correct structure and deactivates Polylang's plugins.
+	 *
+	 * @global $wp_rewrite
+	 *
+	 * @param WP_Site $blog      Blog to set up.
+	 * @param string  $structure Permalinks structure to use.
+	 * @return void
+	 */
+	protected function set_up_blog_without_pll( WP_Site $blog, string $structure ) {
+		global $wp_rewrite;
+
+		switch_to_blog( $blog->blog_id );
+
+		$wp_rewrite->init();
+		$wp_rewrite->extra_rules_top = array();
+		$wp_rewrite->set_permalink_structure( $structure );
+		$wp_rewrite->flush_rules();
+
+		$plugins = get_option( 'active_plugins', array() );
+		update_option( 'active_plugins', array_diff( $plugins, $this->get_plugin_names() ) ); // Ensure Polylang plugins are deactivated.
+
+		restore_current_blog();
+	}
+
+	/**
+	 * Removes *all* filters that could interfere with a test.
+	 * Must be called *before* creating a new Polylang environment in a test.
+	 */
+	protected function clean_up_filters() {
```

Comment:
Why is this necessary? I expect that WordPress will clean up all filters added during the tests.
NB: when I run all 6 tests without the method, only the last one fails.

Reply:
Because of [this](https://github.com/polylang/polylang/blob/a9273e35d2ea81f1487274a4a74837a4f6712e87/tests/phpunit/includes/testcase-multisites.php#L197). If we place a `$this->clean_up_filters();` right after this line, all fine; and if we place it before, it fails.